File: SwiGi.Win/SwiGi.Win/SwiGiEngine.cs
using SwiGi.Win.Hid;

namespace SwiGi.Win;

internal enum EngineStatusKind
{
    Stopped,
    Starting,
    Running,
    Error
}

internal sealed class EngineStatus
{
    public EngineStatusKind Kind { get; init; } = EngineStatusKind.Stopped;
    public string? Keyboard { get; init; }
    public string? Mouse { get; init; }
    public int SwitchCount { get; init; }
    public string? ErrorMessage { get; init; }
}

internal sealed class SwiGiEngine : IDisposable
{
    private readonly object _lock = new();
    private CancellationTokenSource? _cts;
    private Task? _worker;
    private EngineStatus _status = new();

    public event Action<EngineStatus>? StatusChanged;
    public bool VerboseLogging { get; set; }

    public EngineStatus Status
    {
        get { lock (_lock) return _status; }
    }

    public bool IsRunning
    {
        get
        {
            var k = Status.Kind;
            return k is EngineStatusKind.Running or EngineStatusKind.Starting;
        }
    }

    public void Start()
    {
        lock (_lock)
        {
            if (_cts is not null) return;
            _cts = new CancellationTokenSource();
            SetStatus(new EngineStatus { Kind = EngineStatusKind.Starting });
            _worker = Task.Run(() => RunLoopAsync(_cts.Token));
        }
    }

    public void Stop()
    {
        lock (_lock)
        {
            _cts?.Cancel();
            _cts = null;
        }
    }

    public void Dispose()
    {
        Stop();
        try { _worker?.Wait(TimeSpan.FromSeconds(3)); } catch { /* ignore */ }
    }

    private void SetStatus(EngineStatus status)
    {
        lock (_lock) _status = status;
        StatusChanged?.Invoke(status);
    }

    private void Log(string message, bool force = false)
    {
        if (VerboseLogging || force)
            System.Diagnostics.Debug.WriteLine($"[SwiGi] {message}");
    }

    private async Task RunLoopAsync(CancellationToken ct)
    {
        Log("Searching for devices...", force: true);

        DeviceInfo? keyboard;
        DeviceInfo? mouse = null;
        if (keyboard is null)
        {
            SetStatus(new EngineStatus { Kind = EngineStatusKind.Error, ErrorMessage = "Keyboard not found. Check Bluetooth connection." });
            CleanupWorker();
            return;
        }

        mouse = DeviceDiscovery.FindDevice(HidPPConstants.DeviceTypeMouse);
        if (mouse is null)
        {
            keyboard.Dispose();
            SetStatus(new EngineStatus { Kind = EngineStatusKind.Error, ErrorMessage = "Mouse not found. Check Bluetooth connection." });
            CleanupWorker();
            return;
        }

        Log($"Keyboard: {keyboard.Name} (CHANGE_HOST idx={keyboard.ChangeHostIndex})", force: true);
        Log($"Mouse: {mouse.Name} (CHANGE_HOST idx={mouse.ChangeHostIndex})", force: true);

        var totalSwitches = 0;
        var lastResponse = DateTime.UtcNow;
        const int watchdogSeconds = 10;

        SetStatus(new EngineStatus
        {
            Kind = EngineStatusKind.Running,
            Keyboard = keyboard.Name,
            Mouse = mouse.Name,
            SwitchCount = totalSwitches
        });

        try
        {
            while (!ct.IsCancellationRequested)
            {
                if ((DateTime.UtcNow - lastResponse).TotalSeconds > watchdogSeconds)
                {
                    Log($"Watchdog: no response for {watchdogSeconds}s, reconnecting...", force: true);
                    keyboard.Dispose();
                    mouse.Dispose();
                    await Task.Delay(1000, ct);

                    var kb = DeviceDiscovery.FindDevice(HidPPConstants.DeviceTypeKeyboard);
                    if (kb is not null)
                    {
                        keyboard = kb;
                        Log($"Watchdog reconnect: {keyboard.Name}", force: true);
                    }

                    lastResponse = DateTime.UtcNow;
                    continue;
                }

                try
                {
                    keyboard.Transport.Write(HidPPProtocol.Ping.Span);
                }
                catch
                {
                    Log("Keyboard disconnected, waiting for reconnect...", force: true);
                    keyboard.Dispose();

                    DeviceInfo? kbNew = null;
                    for (var attempt = 0; attempt < 120 && !ct.IsCancellationRequested; attempt++)
                    {
                        await Task.Delay(500, ct);
                        kbNew = DeviceDiscovery.FindDevice(HidPPConstants.DeviceTypeKeyboard);
                        if (kbNew is not null) break;
                    }

                    if (kbNew is null)
                    {
                        Log("Keyboard did not return, retrying...", force: true);
                        continue;
                    }

                    keyboard = kbNew;
                    Log($"Keyboard reconnect: {keyboard.Name}", force: true);
                    lastResponse = DateTime.UtcNow;
                    mouse.Dispose();
                    mouse = null;
                    continue;
                }

                var deadline = DateTime.UtcNow.AddMilliseconds(80);
                while (DateTime.UtcNow < deadline && !ct.IsCancellationRequested)
                {
                    byte[]? raw;
                    try
                    {
                        raw = keyboard.Transport.Read(25);
                    }
                    catch
                    {
                        break;
                    }

                    if (raw is null or { Length: < 4 }) continue;
                    if (!HidPPConstants.MsgLengths.TryGetValue(raw[0], out var expectedLen) || raw.Length != expectedLen)
                        continue;

                    var feature = raw[2];
                    var function = raw[3];
                    var swId = (byte)(function & 0x0F);
                    lastResponse = DateTime.UtcNow;

                    if (feature == keyboard.ChangeHostIndex && swId == 0 && raw.Length > 5)
                    {
                        var targetHost = raw[5];
                        Log($"Easy-Switch: {keyboard.Name} → host {targetHost}", force: true);

                        if (mouse is null || !mouse.Transport.IsOpen)
                        {
                            var newMouse = DeviceDiscovery.FindDevice(HidPPConstants.DeviceTypeMouse);
                            if (newMouse is not null)
                                mouse = newMouse;
                            else
                            {
                                Log("Mouse unavailable — will switch on next Easy-Switch", force: true);
                                break;
                            }
                        }

                        try
                        {
                            HidPPProtocol.SendChangeHost(mouse.Transport, HidPPConstants.DevnumberDirect,
                                mouse.ChangeHostIndex, targetHost);
                            totalSwitches++;
                            Log($"CHANGE_HOST → {mouse.Name} → host {targetHost}", force: true);
                            SetStatus(new EngineStatus
                            {
                                Kind = EngineStatusKind.Running,
                                Keyboard = keyboard.Name,
                                Mouse = mouse.Name,
                                SwitchCount = totalSwitches
                            });
                        }
                        catch
                        {
                            Log("CHANGE_HOST to mouse failed, reconnecting mouse...", force: true);
                            mouse.Dispose();
                            await Task.Delay(1000, ct);
                            var newMouse = DeviceDiscovery.FindDevice(HidPPConstants.DeviceTypeMouse);
                            if (newMouse is not null)
                            {
                                mouse = newMouse;
                                try
                                {
                                    HidPPProtocol.SendChangeHost(mouse.Transport, HidPPConstants.DevnumberDirect,
                                        mouse.ChangeHostIndex, targetHost);
                                    totalSwitches++;
                                    Log($"CHANGE_HOST → {mouse.Name} → host {targetHost} (after reconnect)", force: true);
                                    SetStatus(new EngineStatus
                                    {
                                        Kind = EngineStatusKind.Running,
                                        Keyboard = keyboard.Name,
                                        Mouse = mouse.Name,
                                        SwitchCount = totalSwitches
                                    });
                                }
                                catch
                                {
                                    Log("CHANGE_HOST retry failed", force: true);
                                }
                            }
                        }

                        break;
                    }

                    if (swId == 0)
                        Log($"Notification: feat=0x{feature:X2}");
                }

                await Task.Delay(20, ct);
            }
        }
        catch (OperationCanceledException)
        {
            // normal shutdown
        }
        finally
        {
            keyboard.Dispose();
            mouse?.Dispose();
            SetStatus(new EngineStatus { Kind = EngineStatusKind.Stopped });
            CleanupWorker();
        }
    }

    private void CleanupWorker()
    {
        lock (_lock)
        {
            _cts?.Dispose();
            _cts = null;
            _worker = null;
        }
    }
}

File: SwiGi/SwiGi/SwiGiEngine.swift
import Foundation
import os

@MainActor
final class SwiGiEngine: ObservableObject {
    enum Status: Equatable {
        case stopped
        case starting
        case running(keyboard: String, mouse: String, switchCount: Int)
        case error(String)
    }

    @Published private(set) var status: Status = .stopped
    @Published var verboseLogging = false

    var onLog: (@Sendable (String, OSLogType) -> Void)?

    private var workerTask: Task<Void, Never>?
    private let logger = Logger(subsystem: "com.swigi.app", category: "engine")

    var isRunning: Bool {
        if case .running = status { return true }
        if case .starting = status { return true }
        return false
    }

    func start() {
        guard workerTask == nil else { return }
        status = .starting
        workerTask = Task.detached(priority: .userInitiated) { [weak self] in
            await self?.runLoop()
        }
    }

    func stop() {
        workerTask?.cancel()
        workerTask = nil
        status = .stopped
    }

    private func runLoop() async {
        HIDAPIBridge.initialize()

        await updateStatus(.starting)
        log("SwiGi — searching for devices...", level: .info)

        guard var keyboard = DeviceDiscovery.findDevice(wantedType: HIDPPConstants.deviceTypeKeyboard) else {
            await updateStatus(.error("Keyboard not found. Check Bluetooth connection."))
            workerTask = nil
            return
        }

        guard var mouse = DeviceDiscovery.findDevice(wantedType: HIDPPConstants.deviceTypeMouse) else {
            keyboard.close()
            await updateStatus(.error("Mouse not found. Check Bluetooth connection."))
            workerTask = nil
            return
        }

        log("Keyboard: \(keyboard.name) (CHANGE_HOST idx=\(keyboard.changeHostIndex))", level: .info)
        log("Mouse: \(mouse.name) (CHANGE_HOST idx=\(mouse.changeHostIndex))", level: .info)
        log("Ready. Press Easy-Switch on \(keyboard.name).", level: .info)

        var totalSwitches = 0
        var lastResponse = Date()
        let watchdogTimeout: TimeInterval = 10

        await updateStatus(.running(keyboard: keyboard.name, mouse: mouse.name, switchCount: totalSwitches))

        while !Task.isCancelled {
            if Date().timeIntervalSince(lastResponse) > watchdogTimeout {
                log("Watchdog: no response for \(Int(watchdogTimeout))s, reconnecting...", level: .info)
                keyboard.close()
                mouse.close()
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                if let kb = DeviceDiscovery.findDevice(wantedType: HIDPPConstants.deviceTypeKeyboard) {
                    keyboard = kb
                    log("Watchdog reconnect: \(keyboard.name)", level: .info)
                }
                lastResponse = Date()
                continue
            }

            do {
                try keyboard.transport.write(HIDPPProtocol.ping)
            } catch {
                log("Keyboard disconnected, waiting for reconnect...", level: .info)
                keyboard.close()

                var kbNew: DeviceInfo?
                for attempt in 0..<120 where !Task.isCancelled {
                    try? await Task.sleep(nanoseconds: 500_000_000)
                    kbNew = DeviceDiscovery.findDevice(wantedType: HIDPPConstants.deviceTypeKeyboard)
                    if kbNew != nil { break }
                    if attempt % 20 == 19 {
                        log("Reconnect attempt \(attempt + 1)/120...", level: .debug)
                    }
                }

                guard let kbNew else {
                    log("Keyboard did not return, retrying...", level: .default)
                    continue
                }

                keyboard = kbNew
                log("Keyboard reconnect: \(keyboard.name)", level: .info)
                lastResponse = Date()
                mouse.close()
                log("Closed stale mouse transport; will reconnect on next event", level: .debug)
                continue
            }

            let deadline = Date().addingTimeInterval(0.08)
            while Date() < deadline, !Task.isCancelled {
                let raw: Data
                do {
                    guard let data = try keyboard.transport.read(timeout: 25) else { continue }
                    raw = data
                } catch {
                    break
                }

                guard raw.count >= 4 else { continue }
                let reportID = raw[0]
                guard let expectedLen = HIDPPConstants.msgLengths[reportID], raw.count == expectedLen else { continue }

                let feature = raw[2]
                let function = raw[3]
                let swID = function & 0x0F
                lastResponse = Date()

                if feature == keyboard.changeHostIndex, swID == 0, raw.count > 5 {
                    let targetHost = raw[5]
                    log("Easy-Switch: \(keyboard.name) → host \(targetHost)", level: .info)

                    if !mouse.transport.isOpen {
                        log("Mouse transport stale, reconnecting...", level: .debug)
                        if let newMouse = DeviceDiscovery.findDevice(wantedType: HIDPPConstants.deviceTypeMouse) {
                            mouse = newMouse
                        } else {
                            log("Mouse unavailable — will switch on next Easy-Switch", level: .info)
                            break
                        }
                    }

                    do {
                        try HIDPPProtocol.sendChangeHost(
                            transport: mouse.transport,
                            devnumber: HIDPPConstants.devnumberDirect,
                            featureIndex: mouse.changeHostIndex,
                            targetHost: targetHost
                        )
                        totalSwitches += 1
                        log("CHANGE_HOST → \(mouse.name) → host \(targetHost)", level: .info)
                        await updateStatus(.running(keyboard: keyboard.name, mouse: mouse.name, switchCount: totalSwitches))
                    } catch {
                        log("CHANGE_HOST to mouse failed, reconnecting mouse...", level: .default)
                        mouse.close()
                        try? await Task.sleep(nanoseconds: 1_000_000_000)
                        if let newMouse = DeviceDiscovery.findDevice(wantedType: HIDPPConstants.deviceTypeMouse) {
                            mouse = newMouse
                            do {
                                try HIDPPProtocol.sendChangeHost(
                                    transport: mouse.transport,
                                    devnumber: HIDPPConstants.devnumberDirect,
                                    featureIndex: mouse.changeHostIndex,
                                    targetHost: targetHost
                                )
                                totalSwitches += 1
                                log("CHANGE_HOST → \(mouse.name) → host \(targetHost) (after reconnect)", level: .info)
                                await updateStatus(.running(keyboard: keyboard.name, mouse: mouse.name, switchCount: totalSwitches))
                            } catch {
                                log("CHANGE_HOST retry failed — mouse will switch next time", level: .default)
                            }
                        } else {
                            log("Mouse unavailable — will switch on next Easy-Switch", level: .info)
                        }
                    }
                    break
                }

                if swID == 0 {
                    log("Notification: feat=0x\(String(format: "%02X", feature)) [\(raw.prefix(10).map { String(format: "%02x", $0) }.joined())]", level: .debug)
                }
            }

            try? await Task.sleep(nanoseconds: 20_000_000)
        }

        log("Stopping. Total switches: \(totalSwitches)", level: .info)
        keyboard.close()
        mouse.close()
        await updateStatus(.stopped)
        workerTask = nil
    }

    private func updateStatus(_ newStatus: Status) async {
        await MainActor.run {
            status = newStatus
        }
    }

    private func log(_ message: String, level: OSLogType) {
        if level == .debug, !verboseLogging { return }
        logger.log(level: level, "\(message, privacy: .public)")
        onLog?(message, level)
    }
}
